include/jemalloc/internal/thread_event.h

```diff
@@ -0,0 +1,179 @@
+#ifndef JEMALLOC_INTERNAL_THREAD_EVENT_H
+#define JEMALLOC_INTERNAL_THREAD_EVENT_H
+
+#include "jemalloc/internal/malloc_io.h"
+#include "jemalloc/internal/tsd.h"
+
+/*
+ * Maximum threshold on thread_allocated_threshold_fast, so that there is no
+ * need to check overflow in malloc fast path. (The allocation size in malloc
+ * fast path never exceeds SC_LOOKUP_MAXCLASS.)
+ */
+#define THREAD_ALLOCATED_THRESHOLD_FAST_MAX				\
+    (UINT64_MAX - SC_LOOKUP_MAXCLASS + 1)
+
+/*
+ * The maximum wait time should not exceed INT64_MAX, since all the individual
+ * wait times are defined as int64_t.  The maximum wait time should also be at
+ * most UINT64_MAX - SC_LARGE_MAXCLASS + 1, so that whenever an event is
+ * triggered, the accumulated allocation bytes bulk-transferred to the
+ * individual wait times do not overflow uint64_t.
+ *
+ * There is actually one more contraint: SC_LARGE_MAXCLASS - 1 <=
+ * (uint64_t)INT64_MIN (where the uint64_t cast is intentially meant to flip
+ * INT64_MIN); otherwise the accumulated allocation bytes bulk-transferred to
+ * the individual wait times might underflow int64_t.  Luckily this constraint
+ * is automatically guaranteed since SC_LARGE_MAXCLASS is capped by
+ * 2 ** (ptr_bits - 1), which is at most (uint64_t)INT64_MIN.  (See
+ * explanations in sc.h.)
+ */
+#define THREAD_EVENT_WAIT_MAX						\
+    (UINT64_MAX - SC_LARGE_MAXCLASS + 1 < (uint64_t)INT64_MAX ?		\
+     UINT64_MAX - SC_LARGE_MAXCLASS + 1 : (uint64_t)INT64_MAX)
+
+/*
+ * List of all waits, in the following format:
+ *  W(event,		(condition))
+ */
+#define ITERATE_OVER_ALL_WAITS						\
+    W(prof_sample,	(config_prof && opt_prof))
+
+JEMALLOC_ALWAYS_INLINE void
+thread_event_adjust_thresholds_helper(tsd_t *tsd, uint64_t last_event,
+    uint64_t wait) {
+	assert(last_event == tsd_thread_allocated_last_event_get(tsd));
+	assert(wait > 0);
+	uint64_t threshold = last_event + wait;
+	tsd_thread_allocated_threshold_set(tsd, threshold);
+	if (likely(last_event < threshold &&
+	    threshold <= THREAD_ALLOCATED_THRESHOLD_FAST_MAX)) {
+		tsd_thread_allocated_threshold_fast_set(tsd, threshold);
+	} else {
+		/*
+		 * threshold either overflowed or exceeded
+		 * THREAD_ALLOCATED_THRESHOLD_FAST_MAX.  Zero
+		 * thread_allocated_threshold_fast so that
+		 * thread_event_fast() will always trigger.
+		 */
+		tsd_thread_allocated_threshold_fast_set(tsd, 0);
+	}
+}
+
+JEMALLOC_ALWAYS_INLINE bool
+thread_event(tsd_t *tsd, size_t usize) {
```

Comment:
Maybe `thread_event_alloc` instead, to be "namespace conscious"? Also I wonder if we might ever have any deallocation size triggered events? The "event" in the new event handler mainly refers to something that's triggered by the allocation / deallocation, rather than the event *of* allocation / deallocation. So it doesn't feel logically consistent to mix the deallocation logic into this event handler unless we may have some deallocation triggered events.

Reply:
Yes dealloc will trigger tcache GC (currently using ticker on dealloc, which should be replaced by thread.deallocated similarly).

Reply:
Hmm probably not `thread_event_alloc` which sounds like we are allocating for something.

Reply:
Discussed offline. We'll wait until we're at the stage of actually adding the GC support, before trying to decide whether we should split the function into two, or still keep it while letting it accept a pointer to the accumulator, .